src/components/utils/test/thread_test.cc

```diff
@@ -171,10 +292,153 @@ TEST_F(ThreadTest, StartOneThreadTwice_ExpectTheSameThreadStartedTwice) {
+
+  // Wait for the thread to notify it has started execution
+  waitForThreadStart(delegate);
+  EXPECT_TRUE(thread->is_running());
+
+  // Signal the thread to stop and wait for the execution
+  thread->stop();
+  thread->join(Thread::JoinOptionStop::kNoStop);
+  EXPECT_FALSE(thread->is_running());
+
+  // Reset the parameters of the delegate so thread can be started again
+  delegate->notified_ = false;
+  delegate->notify_running_ = false;
+  EXPECT_TRUE(thread->start());
+
+  // Wait for the thread to notify it is running
+  waitForThreadStart(delegate);
+  EXPECT_TRUE(thread->is_running());
+
+  // Make the thread stop and wait for it
+  delegate->signalWakeUp();
+  thread->join(Thread::JoinOptionStop::kNoStop);
+  EXPECT_FALSE(thread->is_running());
+
+  delete delegate;
+  DeleteThread(thread);
+}
+
+TEST(ThreadExecutionFlowTest, StartJoinStart_Success) {
+  // Create a delegate with infinite threadMain
+  InfiniteMainTestDelegate* delegate = new InfiniteMainTestDelegate();
+  Thread* thread = CreateThread(kThreadName.c_str(), delegate);
+  EXPECT_TRUE(thread->start(threads::ThreadOptions()));
+  EXPECT_TRUE(thread->is_joinable());
+
+  // Wait for the thread to notify it has started execution
+  waitForThreadStart(delegate);
+  EXPECT_TRUE(thread->is_running());
+  // Signal the delegate it has to finish the execution of threadMain
+  delegate->notified_ = true;
+  {
+    AutoLock test_lock(test_mutex_two);
+    cond_var_two.Broadcast();
+  }
+  thread->join(Thread::JoinOptionStop::kNoStop);
+  EXPECT_FALSE(thread->is_running());
+
+  // Reset the parameters of the delegate so thread can be started again
+  delegate->notified_ = false;
+  delegate->notify_running_ = false;
+
+  EXPECT_TRUE(thread->start());
+
+  // Wait for the thread to notify it has started execution
+  waitForThreadStart(delegate);
+  EXPECT_TRUE(thread->is_running());
+
+  // Signal the delegate it has to finish the execution of threadMain
+  delegate->signalWakeUp();
+  thread->join(Thread::JoinOptionStop::kNoStop);
+  EXPECT_FALSE(thread->is_running());
+
+  delete delegate;
+  DeleteThread(thread);
+}
+
+TEST(ThreadExecutionFlowTest, DeleteJoinable_Success) {
+  // Create a delegate with delayed exit of threadMain
+  DelayedExitTestDelegate* delegate = new DelayedExitTestDelegate();
+  Thread* thread = CreateThread(kThreadName.c_str(), delegate);
+  EXPECT_TRUE(thread->start(threads::ThreadOptions()));
+  EXPECT_TRUE(thread->is_joinable());
+
+  // Wait for the thread to notify it has started execution
+  waitForThreadStart(delegate);
+
+  // Reset the parameters of the delegate
+  delegate->notify_running_ = false;
+  EXPECT_TRUE(thread->is_running());
+  DeleteThread(thread);
+
+  // Wait for threadMain to notify it is exiting
+  waitForThreadStart(delegate);
+  EXPECT_TRUE(delegate->notified_exit_);
+
+  // Creating a new thread so the delegate can be deleted
+  Thread* thread2 = CreateThread(kThreadName.c_str(), delegate);
```

Comment:
@pvvasilev, please rename `thread2`. And for future, use numbers for naming it's bad practise.